### PR TITLE
[FSSS-168] Feat: lazy loading and improvements (CLS)

### PR DESCRIPTION
**Pull request by eduardoformiga**
## What's the purpose of this pull request?
This PR aims to lazy loading either the components below the fold or the components that delay showing so that we have a fallback loading component. To accomplish that, we are using the [loadable-components](https://github.com/gregberge/loadable-components) and also the `lazy/Suspense` from `react`.

Along the way, we notice that the `Cumulative Layout Shift (CLS)` has an influence on the `lazy loading`/`code-splitting` on the components. So, we addressed this Core Web Vital metric and now we have almost the **best** score on all the pages (Homepage, PLP, PDP, and Search Page) so far (Desktop and mobile).

This PR also improves the User Experience (UX) to prevent some Layout shift and blink components in specific cases like when we change the Filter/Sort behavior then some components are being blinked.

It also moves some Styles to the corresponding pages/components.

Adequates the Search page to use the new structure.

- [x] Lazy Loading Layout and PLP components.
- [x] Solve CLS on PLP.
- [x] Solve CLS on Search Page.
- [x] Solve CLS on Homepage.
- [x] Improves UX. CLS and rerender (blink).
- [x] Attempts to improve the TTI (Time To Interactive).
- [x] Moves some Styles to the corresponding pages/components.
- [x] Adequates the Search page to use the new structure.

||Before|After|
|-|-|-|
|CLS on PLP (Desktop)|Screen Shot 2022-02-02 at 17 25 38|Screen Shot 2022-02-02 at 17 25 30|
|CLS on PLP initial loading|yO7JwMDdSt|5WrLrdhO9c|
|UX after changing filter/sorting|2OKSNzUxaa|VvtijUAeVl|
|CLS on Search Page (Desktop)|Screen Shot 2022-02-03 at 14 26 15|Screen Shot 2022-02-03 at 14 32 12|
|CLS on Search Page initial loading|beunJQHGP8|th6b4vnHt6|
|CLS on Homepage (Desktop)|Screen Shot 2022-02-03 at 16 32 52|Screen Shot 2022-02-03 at 16 33 01|

[According to Gatsby](https://www.gatsbyjs.com/docs/how-to/performance/improving-site-performance/#step-5-on-critical-paths-lazy-load-below-the-fold-components):

> loadable-components is the recommended lazy-loading solution for all server-side-rendered React applications, including Gatsby websites.

So First I tried to use the native lazy/Suspense, but either the build failed or the application broke someplace. Thus, I switch to using the loadable, probably because there is some Server Side rendered dependency and it worked as aspected.

Finally, for that pages that use both, lazy and loadable, I kept the preference to use only loadable for the sake of avoiding loading 2 different things for the same purpose.

## How to test it?
https://preview-300--base.preview.vtex.app/office/

## References
https://pt-br.reactjs.org/docs/code-splitting.html#reactlazy
https://www.gatsbyjs.com/docs/how-to/performance/improving-site-performance/#step-5-on-critical-paths-lazy-load-below-the-fold-components
https://github.com/gregberge/loadable-components
https://web.dev/interactive/
https://web.dev/i18n/pt/cls/

### CODE REVIEW

**Review by ArthurTriis1 — APPROVED** [review]

File: src/components/sections/ProductGallery/ProductGallery.tsx [hunk 1]
```diff
@@ -24,144 +25,140 @@ function ProductGallery({ title, slug }: Props) {
...
-    return <div className="product-listing__data-loading">Loading…</div>
-  }
-
   return (
-    <div className="product-listing / grid-content-full">
-      <div className="product-listing__content-grid / grid-content">
-        <div className="product-listing__filters">
+    <>
+      <div className="product-listing__filters">
+        <Suspense fallback={null}>
```
Comment (danzanzini) [inline src/components/sections/ProductGallery/ProductGallery.tsx:40]: Just a question: Why did you change to use loadable at the Layout and to use Lazy here?

**Review by saranicoly — APPROVED** [review]

File: src/components/sections/ProductGallery/ProductGallery.tsx (hunk 1 above)
Comment (eduardoformiga) [reply src/components/sections/ProductGallery/ProductGallery.tsx:42]: I prefer to use the native `lazy`/`Suspense` whenever possible. 
But since we need to use `loadable` in some components of our `Layout` because if we didn't the build fails. There we switch to loadable (that deals better with Server Side).
What do you think?

File: src/components/sections/ProductGallery/product-gallery.scss [hunk 2]
```diff
@@ -85,8 +58,11 @@
 
 .product-listing__results {
   --padding: var(--space-1);
+  --product-listing-grid-height: 2197px;
```
Comment (hellofanny) [inline src/components/sections/ProductGallery/product-gallery.scss:61]: I guess this is also temporary right?

File: src/styles/pages/search.scss [hunk 3]
```diff
@@ -0,0 +1,31 @@
+@import "../vendors/include-media.scss";
+
+.product-listing {
```
Comment (hellofanny) [inline src/styles/pages/search.scss:3]: Maybe, we can find a way to extract this common block of style for both pages. 🤔

**Review by hellofanny — APPROVED** [review]
LOVE IT 🤩🤩🤩🤩🤩! It's a very good improvement!! Well done 🥳

File: src/components/sections/ProductGallery/ProductGallery.tsx (hunk 1 above)
Comment (hellofanny) [reply src/components/sections/ProductGallery/ProductGallery.tsx:42]: I think this is an nice thing to be documented, but not sure where 🙈  (this kind of decision)

File: src/components/sections/ProductGallery/product-gallery.scss (hunk 2 above)
Comment (eduardoformiga) [reply src/components/sections/ProductGallery/product-gallery.scss:61]: I got this number testing in different screen sizes and it is the minimum when we have the page loaded with the products. If we ensure that the content that will fill in the loading space (skeletons?) has at least this height we can remove it afterward. this height ensures good CLS so far.

File: src/styles/pages/search.scss (hunk 3 above)
Comment (eduardoformiga) [reply src/styles/pages/search.scss:3]: Yeahh, I was thinking about that but I didn't find a good solution so far 🤔 
Suggestions?

File: src/components/sections/ProductGallery/ProductGallery.tsx (hunk 1 above)
Comment (filipewl) [reply src/components/sections/ProductGallery/ProductGallery.tsx:42]: Maybe it's because I don't know the difference between `lazy` and `loadable`, but I also like the idea of documenting when to use each.
🙈

**Review by filipewl — APPROVED** [review]
Very nice improvements to core web vitals! 👏

File: src/components/sections/ProductGallery/ProductGallery.tsx (hunk 1 above)
Comment (eduardoformiga) [reply src/components/sections/ProductGallery/ProductGallery.tsx:42]: [According to Gatsby](https://www.gatsbyjs.com/docs/how-to/performance/improving-site-performance/#step-5-on-critical-paths-lazy-load-below-the-fold-components):

> loadable-components is the recommended lazy-loading solution for all server-side-rendered React applications, including Gatsby websites.

So First I tried to use the native `lazy`/`Suspense`, but either the build failed or the application broke someplace. Thus, I switch to using the `loadable`, probably because there is some Server Side rendered dependency and it worked as aspected.

Finally, for that pages that use both, `lazy` and `loadable`, I kept the preference to use only `loadable` for the sake of avoiding loading 2 different things for the same purpose.

Where do you think is the best place to documented it?

File: src/components/sections/ProductGallery/ProductGallery.tsx [hunk 4]
```diff
@@ -26,144 +27,140 @@ function ProductGallery({ title, slug }: Props) {
...
+          />
+        </Suspense>
+      </div>
+
+      <div
+        className="product-listing__results-count"
+        data-testid="total-product-count"
+        data-count={totalCount}
+      >
+        {<h2>{totalCount} Results</h2>}
```
[the remark this answers is not in the document]
Comment (eduardoformiga) [reply src/components/sections/ProductGallery/ProductGallery.tsx:57]: probably not hahaha, thank you!

File: src/components/sections/ProductGallery/ProductGallery.tsx (hunk 1 above)
Comment (filipewl) [reply src/components/sections/ProductGallery/ProductGallery.tsx:42]: Since we seem to only have the README as the documentation, maybe it could be added under the **⚡ Performance & QA** section. At the same time, I'm not sure it's something we need to add there.
😕

Comment (hellofanny) [reply src/components/sections/ProductGallery/ProductGallery.tsx:42]: How about a comment in the code? (Like we have on `Sentinel` component) Or doesn't need it?

File: src/components/sections/ProductGallery/product-gallery.scss (hunk 2 above)
Comment (hellofanny) [reply src/components/sections/ProductGallery/product-gallery.scss:61]: Oh great! How about a ticket on Jira for follow-up/check. Or a comment here? //Check it after apply Skeleton.

Comment (eduardoformiga) [reply src/components/sections/ProductGallery/product-gallery.scss:62]: Nice, Done it here 9a7948dbb2f8b9828ef146d7e49e5b9185e9b8e9

File: src/styles/pages/search.scss (hunk 3 above)
Comment (eduardoformiga) [reply src/styles/pages/search.scss]: I paired with @renatamottam and we tried some solutions.  So far we extract the common block into a `ProductListing` component and reuse it on both pages Search and PLP.

File: src/components/sections/ProductGallery/ProductGallery.tsx (hunk 1 above)
Comment (eduardoformiga) [reply src/components/sections/ProductGallery/ProductGallery.tsx:42]: I added this section on our `README.md` here 9bda2b62b32ae248059d6e0a7f39d9584ec72665.
Hmm, I think if we added the comment in the code we'll have a lot of comments, and maybe pollute the pages, for all places where we use lazy loading. 
What's are your opinions?

File: README.md [hunk 5]
```diff
@@ -403,6 +403,16 @@ Looking for more guidance? Full documentation for Faststore lives [on this GitHu
 
 This project has strict performance budgets. Right out of the box, this project performs around 95 on Google's Page Speed Insights website, which usually is way more strict than your laptop's chrome lighthouse. Every time you commit to the repository, our QA bots will run and evaluate your code quality. We recommend you NEVER put in production a code that breaks any of the bots. If a bot breaks and still you need to put the code into production, change the bot config (`lighthouserc.js`, `cypress.json`) to make it pass and merge. This way you ensure your website will keep performing well during the years to come.
 
+### Lazy loading components
+
+[According to Gatsby](https://www.gatsbyjs.com/docs/how-to/performance/improving-site-performance/#step-5-on-critical-paths-lazy-load-below-the-fold-components):
+
+> `loadable-components` is the recommended lazy-loading solution for all server-side-rendered React applications, including Gatsby websites.
+
+So First, Try to use the native `lazy`/`Suspense` alternative. But if there is some Server Side Rendered dependency, switch to using the `loadable-components`.
```
Comment (hellofanny) [inline README.md:412]:
```suggestion
So First, try to use the native `lazy`/`Suspense` alternative. But if there is some Server Side Rendered dependency, switch to using the `loadable-components`.
```

File: README.md [hunk 6]
```diff
@@ -403,6 +403,16 @@ Looking for more guidance? Full documentation for Faststore lives [on this GitHu
...
+
+Finally, for that pages that can use both `lazy` and `loadable`, let's keep the preference to use only `loadable` for the sake of avoiding loading two different things for the same purpose.
```
Comment (hellofanny) [inline README.md:414]:
```suggestion
Finally, for that pages that can use both `lazy` and `loadable`, keep the preference to use only `loadable` for the sake of avoiding loading two different things for the same purpose.
```

File: src/components/sections/ProductGallery/ProductGallery.tsx (hunk 1 above)
Comment (hellofanny) [reply src/components/sections/ProductGallery/ProductGallery.tsx:42]: Wonderful! 🌟
